distributed/worker.py

```diff
@@ -1712,11 +1714,19 @@ async def close(
             for executor in self.executors.values():
                 if executor is utils._offload_executor:
                     continue  # Never shutdown the offload executor
-                if isinstance(executor, ThreadPoolExecutor):
-                    executor._work_queue.queue.clear()
-                    executor.shutdown(wait=executor_wait, timeout=timeout)
-                else:
-                    executor.shutdown(wait=executor_wait)
+
+                def _close():
+                    if isinstance(executor, ThreadPoolExecutor):
+                        executor._work_queue.queue.clear()
```

Comment:
is this the same as cancel_futures=True on python3.9?

https://github.com/python/cpython/blob/3.9/Lib/concurrent/futures/thread.py#L219-L228

Reply:
clearing the queue without cancelling the futures will cause any tasks waiting on `await run_in_executor(executor, ...)` to hang

Reply:
This was added in https://github.com/dask/distributed/pull/2905 / https://github.com/dask/distributed/commit/051a79e05c501dcea57a19715f59d0cd203f6be6

Not a lot of additional context but I assume the intention is as you are describing to cancel queued up work.

---

This only affects our vendored threadpool which is used in two places

* TCP https://github.com/dask/distributed/blob/be4fc7f7247ef6244f9a48e36b9fb2d1d4a4b21a/distributed/comm/tcp.py#L394-L396
* Worker  https://github.com/dask/distributed/blob/be4fc7f7247ef6244f9a48e36b9fb2d1d4a4b21a/distributed/worker.py#L952-L967

The TCP pool is not even closed in this shutdown procedure so it should not be relevant, although I think it is a mistake to use the vendored pool there.

The queue on the worker threadpool is only filled if we know there is in fact capacity available so this is only a problem if we're shutting down just after a task was transitioned to executing, before a threadpool worker could pick up the task.... probably very unlikely but still something we should fix.

Good catch!

Regarding a fix, I guess the "clean" way would be to update our vendored version of the threadpool to properly support the cancel_futures kwarg. I'll definitely not do this in this PR.

However, I guess we should remove this line of code immediately since it either has no effect or can cause a deadlock. Is this correct?

Reply:
yep that makes sense!